### PR TITLE
GCC port and new defines for new ARM port

### PR DESCRIPTION
**Pull request by a1batross**
A working port to GCC 7.3. 

I also have added two new defines:
* HAVE_SSE enabled automatically when compiler is enabled SSE support. It will be used for updated ARM port of ReGameDLL_CS.
* NO_AMXX_HACKS disables ASM inline. It will be used for updated ARM port of ReGameDLL_CS.

### CODE REVIEW

**Comment by a1batross** [comment]
As I am afraid to do any changes in Gradle, because I don't know it, I've published a CMakeLists.txt used to compile this to Gist here:
https://gist.github.com/a1batross/3b4e1ecc34c7a29646b6acc05a8d90f5

File: regamedll/dlls/saverestore.cpp [hunk 1]
```diff
@@ -28,6 +28,11 @@
 
 #include "precompiled.h"
 
+// this should be defined(otherwise link will fail), but MSVC doesn't require it
+#ifndef _MSC_VER
+constexpr int CSaveRestoreBuffer::m_Sizes[];
```
Comment (WPMGPRoSToTeMa) [inline regamedll/dlls/saverestore.cpp]: I'm not sure, but looks like expression `m_Sizes[x]` (where `x` isn't constexpr) requires address evaluation and then this definition is required.

File: regamedll/dlls/util.h [hunk 2]
```diff
@@ -323,7 +323,8 @@ inline T *UTIL_FindEntityInSphere(T *pStartEntity, const Vector &vecCenter, floa
 	pentEntity = FIND_ENTITY_IN_SPHERE(pentEntity, vecCenter, flRadius);
 	if (!FNullEnt(pentEntity))
 	{
-		return (T *)CBaseEntity::Instance(pentEntity);
+		// a1ba: can't use CBaseEntity::Instance here, because circular dependency in cbase.h
```
Comment (WPMGPRoSToTeMa) [inline regamedll/dlls/util.h]: Can you give more info about this?

**Comment by WPMGPRoSToTeMa** [comment]
> A working port to GCC 7.3.

Does it work on GCC 8.1?

File: regamedll/dlls/saverestore.cpp (hunk 1 above)
Comment (WPMGPRoSToTeMa) [reply regamedll/dlls/saverestore.cpp]: This was fixed in C++17.

**Comment by a1batross** [comment]
Yes, it will should work for any version of GCC supporting C++14. 

But I did not tried it. Unfortunately, GCC 8.1 is not yet arrived even in Ubuntu 18.04. :)

File: regamedll/dlls/util.h (hunk 2 above)
Comment (a1batross) [reply regamedll/dlls/util.h]: In cbase.h an util.h is included before CBaseEntity definition. So after preprocessing it's look like:
```
a template function with CBaseEntity::Instance call
...
...
...
actual CBaseEntity::Instance definition
```

In fact, Instance() method performs a nullptr check and returns private data of specified edict or world edict. 

This function already does a null check, so just GET_PRIVATE is enough.

File: regamedll/dlls/saverestore.cpp (hunk 1 above)
Comment (a1batross) [reply regamedll/dlls/saverestore.cpp]: Ah, I see. Thanks for pointing me. :)

**Pull request merged**